src/libsyntax/ext/expand.rs

```diff
@@ -695,6 +737,30 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         }
     }
 
+    fn gate_proc_macro_expansion_kind(&self, span: Span, kind: ExpansionKind) {
+        let kind = match kind {
+            ExpansionKind::Expr => "expressions",
+            ExpansionKind::OptExpr => "expressions",
+            ExpansionKind::Pat => "patterns",
+            ExpansionKind::Ty => "types",
+            ExpansionKind::Stmts => "statements",
+            ExpansionKind::Items => return,
+            ExpansionKind::TraitItems => return,
+            ExpansionKind::ImplItems => return,
+            ExpansionKind::ForeignItems => return,
```

Comment:
What happens if a proc macro attribute is applied to a `match` arm (`match 0 { #[attr] _ => {} }`) or a generic parameter (`fn f<#[attr] T>`)?

Reply:
I suspect it already produces some kind of error, but it would be good to have a test.